src/security/client_ip.cpp

```diff
@@ -0,0 +1,601 @@
+#include "client_ip.h"
+
+#include <array>
+#include <string_view>
+
+#include "util.h"
+
+extern "C" {
+#include <arpa/inet.h>
+#include <ngx_config.h>
+#include <ngx_hash.h>
+#include <ngx_http_request.h>
+#include <ngx_list.h>
+}
+
+using namespace std::literals;
+namespace dnsec = datadog::nginx::security;
+using datadog::nginx::to_string_view;
+
+namespace {
+
+struct IpAddr {
+  int af;
+  union {
+    struct in_addr v4;
+    struct in6_addr v6;
+  } u;
+
+  bool empty() const { return af == 0; }
+  bool is_ipv4() const { return af == AF_INET; }
+  bool is_ipv6() const { return af == AF_INET6; }
+
+  std::optional<std::string> to_string() const {
+    char buf[INET6_ADDRSTRLEN];
+    if (inet_ntop(af, reinterpret_cast<const char *>(&this->u), buf,
+                  sizeof(buf)) == nullptr) {
+      return std::nullopt;
+    }
+    return {buf};
+  };
+
+  bool is_private() const {
+    if (af == AF_INET) {
+      return is_private_v4();
+    } else {
+      return is_private_v6();
+    }
+  }
+
+  bool is_private_v4() const;
+  bool is_private_v6() const;
+
+  static std::optional<IpAddr> from_string(std::string_view addr_sv,
+                                           int af_hint = 0);
+};
+
+std::optional<IpAddr> IpAddr::from_string(std::string_view addr_sv,
+                                          int af_hint) {
+  IpAddr out;
+  std::string input_nulterminated{addr_sv};
+  auto *addr_nult = input_nulterminated.c_str();
+
+  if (af_hint != AF_INET6) {
+    int ret = inet_pton(AF_INET, addr_nult, &out.u.v4);
+    if (ret == 1) {
+      out.af = AF_INET;
+      return {out};
+    }
+  }
```

Comment:
Is it intentional to ignore the hint if `inet_pton` fails? The logic then differs when the hint is incorrect for IPv6

Reply:
if it fails, it might still be an ipv6-mapped ipv4 address, which it why it continues. However, from the usage isn't very accurate, because it calls with AF_INET6 if it within brackets. For that what matters is that it's formally ipv6. I'll adjust.